[Bat] 미션 5: 보드에 위치 부여 및 점수 계산

kjk402 opened the pull request:
안녕하세요, 웹 백엔드과정 Bat입니다.

하기와 같이 미션5 구현하여 pull request 드립니다.

• mission5 summary
=====================================
• enum 적용
미션 4까지 상수로 선언되어 있던 Color와 Representation을 enum 적용하였습니다.

• createWhite(), createBlack()
팩토리 메소드에서 중복을 제거하기 위해 생성하였습니다. 

• Position class
체스판에서 기물들의 위치를 나타내는 Position을 생성하였습니다.

• piecesSort : 포인트가 높은 기물 순서로 정렬 메소드 
compareTo를 포인트로 비교하도록 정의하였습니다. 

• 리스트 선회 방식을 사용합니다.
이번 미션에서의 대부분의 요구사항은 리스트 선회방식으로 구현하였습니다. 
해당 방식을 많이 사용하여 요구사항을 구현하다 보니 겹치는 코드가 있습니다. 
하지만 각각의 목적이 조금씩 달라 중복제거를 하지 못하였습니다. 

=====================================

이번 미션을 진행하면서 이전까지 구현해 놓은 코드의 구조를 불가피하게 많이 수정해야 했습니다. 
그러하여 난이도가 다소 느껴져 코딩하는 데 있어 시간이 오래 걸렸으며, 클린 코딩을 깊이 신경 쓰지 못하였습니다. 

부족한 코드 양해 부탁드립니다.

귀하신 시간 내주어 리뷰해 주셔서 감사드립니다.

kjk402 committed:
xml file

kjk402 committed:
Color 와Representation enum 적용

상수로 선언되어있던 Color 와 Representation을 enum으로 변경.

kjk402 committed:
gitignore

kjk402 committed:
Fix PieceTest class

kjk402 committed:
Add method createWhite, createBlack in Piece class

팩토리 메소드 리팩토리를 위해 메소드 추가.

kjk402 committed:
Fix factory method

Piece class에 있는 팩토리 메서드들을 createWhite와 createBlack을
사용하여 중복을 제거하였습니다.

kjk402 committed:
Create Rank class for initialize pieces

kjk402 committed:
add setPieces in Rank clas

Rank class에서 blank 를 포함하여 모든 기물들을 List 하나에 추가 한다.

kjk402 committed:
Piece equals 정의

piece의 Type 과 color 가 같다면 동일한 기물로 판정하는 equals 정의

kjk402 committed:
Create Position class, add position in Piece class

기물의 위치를 뜻하는 position class를 생성하고 Piece class에 관련 변수를
추가하였습니다.

kjk402 committed:
Add Method getFile, getRank in Position class

Position의 rank값과 file 값을 불러 올 수 있도록 get메소드
생성하였습니다.
추가로 Rank class 와 Piece class에서도 Position을 불러오는 메소드
생성하였습니다.

kjk402 committed:
Board 출력 수정.

Rank 값이 같은 기물들만 문자열 반환하는 getRank()메서드 생성하였습니다.
getRank를 BOARD_SIZE만큼 반복하도록 하여 Board 전체를 문자열로
반환합니다.

kjk402 committed:
Create findPiece in BoardTest class

보드의위치로 기물을 조회하는 findPiece 메소드 생성.

kjk402 committed:
특정 위치에 기물을 배치하는 메소드 move 생성.

kjk402 committed:
Create move test in BoardTest Class

move 메소드의 정상동작 확인하기 위해 BoardTest class에 테스트 생성.

kjk402 committed:
기물의 point 부여

kjk402 committed:
같은 color를 가진 기물들 끼리 point 합산하는 caculcatePoint 메소드 생성.

caculcatePoint 메소드는 List에 있는 기물들 중 color가 같은
기물들의 point를 가져와 합산합니다.

kjk402 committed:
불필요한 메서드 제거.

kjk402 committed:
기물들의 정렬을 위해 piecesSort 생성과 compateTo 메소드 정의.

kjk402 committed:
piecesSort 메서드의 정상 동작확인을 위해 piecesSortByPoint 테스트 생성.

kjk402 committed:
Fix PieceTest class

kjk402 committed:
Main class에 불필요한 코드 제거

main에서 test 용도로 사용했던 코드 삭제했습니다.

kjk402 committed:
같은 기물을 개수를 세는 메서드 재생성.

해당 메서드 사용되지 않는 메서드로 오인하여 삭제했었습니다.
메서드 재생성하여 BoartTest 에서 정상적으로 테스트하였습니다.

kjk402 committed:
코드 정렬.

kjk402 committed:
Position class 에서 불필요한 코드 삭제.

ksundong reviewed (commented):
`gitignore` 변경하시는 과정에서 버전관리가 필요없는 파일들이 대거 등록되었네요.
이런 부분은 꼭 신경써주셔야합니다!

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
-        }
-    }
+    public static final int BOARD_SIZE = 8;
+    private final ArrayList<Rank> rankList = new ArrayList<>();
```
이 부분은 왜 이전 코드에선 `List`를 사용하셨다가 `ArrayList`가 되었을까요?

kjk402 replied:
요구사항 4의 보드 자료구조 개선 방법 힌트에서
ArrayList<Rank> 구조로 변경하는 것을 고려해 본다.
라는 문구를 잘못이해한 것같습니다.;; ㅎㅎ 
지금은 다시 List<Rank> 로 변경하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
+    public void initialize() {
+        rankList.add(Rank.setWhitePieces(0));
+        rankList.add(Rank.setWhitePawns(1));
+        rankList.add(Rank.setBlackLine(2));
```
`BlackLine`이 맞나요?

kjk402 replied:
BlankLine이 맞습니다. ㅎㅎ 자동 완성으로 오타 생긴 것 같습니다.
수정하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
         for (int i = 0; i < BOARD_SIZE; i++) {
-            addWhitePawn(new Piece(Piece.WHITE, Piece.WHITE_PAWN_REPRESENTATION));
-            addBlackPawn(new Piece(Piece.BLACK, Piece.BLACK_PAWN_REPRESENTATION));
+            boardOutput.append(appendNewLine(getRank(rankList.get(BOARD_SIZE - 1 - i)) + " " + (BOARD_SIZE - i)));
```
`BOARD_SIZE - i` 가 중복되지 않나요?

kjk402 replied:
6b810c5 에서 중복제거하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
-        addBlackPiece(Piece.createBlackRook());
+    public Piece findPiece(String position) {
+        Position tempPosition = new Position(position);
+        return rankList.get((tempPosition.getFile())).getPiece(tempPosition.getRank());
```
`tempPosition.getFile()`은 불필요한 괄호가 들어간 것 같은데, 제거가 불가능할까요?

kjk402 replied:
6b810c5 에서 괄호 제거해주었습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
-        addBlackPiece(Piece.createBlackKnight());
-        addBlackPiece(Piece.createBlackRook());
+    public Piece findPiece(String position) {
+        Position tempPosition = new Position(position);
```
`temp`보다 더 좋은 네이밍은 없을까요? `target`이라던가... 더 좋은 네이밍을 한 번 고민해봐주세요!

kjk402 replied:
네 제 생각에도 temp는 맞지 않는 것같습니다.
6b810c5 에서 targetPosition으로 수정하였습니다.
좋은 의견 감사합니다.

ksundong replied:
바뀐거 맞나요? 확인 해주세요~

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
+    public void move(String position, Piece piece) {
+        Position tempPosition = new Position(position);
+        rankList.get(tempPosition.getFile()).move(tempPosition.getRank(), piece);
+        Rank rank = new Rank();
+        rank.addPiece(piece);
```
이 부분은 어떤 이유에 의해서 추가된 코드인가요? 동작에는 아무런 영향을 미치지 않을 것 같은데요!

kjk402 replied:
move메서드 처음 생성 시에 동작 확인용으로 넣었습니다. 해당 코드 삭제처리하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
 
-    private String getWhitePiecesResult() {
-        return getPiecesResult(whitePiecesList);
+    public String piecesSort(List<Piece> list, Color color) {
```
`list`라는 이름보단 `pieceList`와 같이 좀 더 특정하기 쉬운 이름을 작성해주세요!

kjk402 replied:
해당 피드백 역시 6b810c5 에서 pieceList로 변경하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
 
-    private String getBlackPiecesResult() {
-        return getPiecesResult(blackPiecesList);
+    private double getPoint(Rank rank, Color color) {
+        double sum = 0.0;
+        for (Piece piece : rank.getPieceList()) {
+            if (piece.getColor() == color) {
+                sum += piece.getType().getDefaultPoint();
+            }
+        }
+        return sum;
     }
```
이게 `Board`가 가지고 있어야할 메서드일까요?
테스트를 수행할 수 있도록 변경해보세요!

kjk402 replied:
Board 클래스에서 한번에 계산하고 싶어서 이렇게 코드를 짜봤습니다. 그러하여 Rank class에서 getPieceList라는 메소드를 생성하여 Board에서 리스트를 불러와 계산합니다.

ksundong replied:
board에서 한 번에 하는 것과, 책임을 분리하는 것은 다릅니다.
`Rank`에서 `getPoint`를 하는 방향은 어떨까요?

kjk402 replied:
5c3a4cb 에서 Rank 클래스로 메소드 이동시켰습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
     }
 
-    private String getPiecesResult(List<Piece> pieces) {
+    private String getRank(Rank rank) {
```
네이밍이 조금 와닿지가 않네요. 그리고, 역시 이것도 `Board`가 가지고 있어야 할 메서드일까요?

kjk402 replied:
위 피드백 답변과 동일한 생각으로 코딩하였습니다.

죄송하지만 Rank 클래스에서 계산후 값을 Board 클래스에 전달하는 것이 제가 한 방식과 어떠한 큰 차이가 있는 지 알려주시면 감사하겠습니다.

메소드 네이밍 getRepresentationByRank 으로 변경하였습니다.

ksundong replied:
객체지향이라는 원칙에 따라 생각하시면 됩니다.
객체지향은 그냥 메서드만 쓰는 것이 아니라, 각 객체별 역할과 책임을 분담시켜서 객체간의 협력을 통해 문제를 해결하는 것이 객체지향 프로그래밍입니다.
단순히 Board에서 모든 일을 다 처리할 것이라면, 굳이 클래스를 나누는 의미가 퇴색될 수도 있습니다. 이런 부분을 염두에 두신다면, 좋을 것 같아요.

사실 이 메서드가 하는 일은
1. `Rank`안에 존재하는 `Piece` 들을 꺼내와서 (`Board`가 `Piece`의 존재를 알아야 할까요?)
2. `Piece`의 `representation`을 합쳐서
3. `Rank`의 `representation`을 만들어 반환한다.
인데요. 과연 이게 `Board`의 책임인지를 생각해보시면 좋을 것 같습니다.

---

그리고 `get`, `set` 네이밍은 최대한 지양하시는 것을 추천드리고 싶습니다.
여기서는 하는 동작이 `get` 해오는게 맞는데, 조금 그런 코드가 자주 나와서 경계하시라는 말씀 드리고 싶어요.

kjk402 replied:
네 Rank에서 해주는 일이 맞다고 생각하여 메소드 옮겨주었습니다.

get 과 set을 메서드 명에 사용하는 것을 자제하도록 하겠습니다.

감사합니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,89 @@
+        return pieceList.get(rankIndex);
+    }
+
+    public static Rank setWhitePieces(int index) {
```
`set` 같은 네이밍 보다는 `create`를 사용해주는 것이 더 나았을 것 같아요.

`createWhitePiecesRank` 정도로요.

kjk402 replied:
e16b48d - 메소드명에서 set이 사용되는 메소드들의 네이밍 변경하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,89 @@
+        return rank;
+    }
+
+    public static Rank setBlackLine(int index) {
```
`setBlankLine`이 맞겠죠?

ksundong replied:
그리고 이렇게 Blank 부분을 채워주면 어떤 장점이 있을까요?

kjk402 replied:
반복문을 사용해서 중복을 줄여준다는 것외에는 잘 모르겠습니다...

ksundong replied:
`null`일 때와 그렇지 않을 때를 한 번 고민해보시는걸 추천드립니다!

kjk402 replied:
아하 null로 두는 것이아니라 NOCOLOR 와 NO_PIECE 로 관리하는 것에대해 여쭤보신 것이네요. 전 메소드 사용 이유 여쭤보신 줄 알았습니다.  ㅎㅎ

제가 생각했을 때는 체스판에서 모든 위치가 객체를 가지고 있어야 게임을 하였을 때 관리가 편리해서인 것 같습니다.

예를 들어 기물을 움직일때 targetPosition 에 기물이 없거나 다른 color를 가진 기물이 있어야 해당 위치로 움직일 수 있도록 검증하는 단계가 필요한데 이 때 NO_PIECE가 사용되기 때문입니다.

ksundong commented on src/main/java/net/kjk402/chess/pieces/Piece.java:
```diff
@@ -1,95 +1,163 @@
+        }
+    }
+
+    private Position position;
```
저라면 `position`도 `final`로 만들어줬을 것 같아요.

kjk402 replied:
7a6d138 에서 final 로 선언하였습니다.

ksundong replied:
음 별 다른 문제는 발생하지 않던가요?

kjk402 replied:
문제 발생해서 원복하였습니다.;;;

ksundong replied:
`final`을 붙이고 사용할 수 있는 방법에 대해서 고민해보세요~

ksundong commented on src/main/java/net/kjk402/chess/pieces/Piece.java:
```diff
@@ -1,95 +1,163 @@
+    }
+
+    private Position position;
+
+    private final Color color;
+
+    private final Type type;
```
왠만하면 멤버변수들 사이는 붙여주시는걸 추천합니다~

kjk402 replied:
7a6d138 에서 멤버변수들 사이 붙여주었습니다.

ksundong commented on src/main/java/net/kjk402/chess/pieces/Piece.java:
```diff
@@ -1,95 +1,163 @@
 
     public boolean isPawn() {
-        return getRepresentation() == WHITE_PAWN_REPRESENTATION || getRepresentation() == BLACK_PAWN_REPRESENTATION;
+        return getRepresentation() == Type.PAWN.getBlackRepresentation() || getRepresentation() == Type.PAWN.getWhiteRepresentation();
```
Type가지고 충분히 판단할 수 있지 않을까요?

kjk402 replied:
네 Type에서 확인가능합니다.
미션4에서 만든 isPawn 메소드는 미션4 구조상 getRepresentation을 사용하여 일일히 비교했으나, 이제는 enum의 Type만으로 간단히 확인가능하여 해당 메소드 삭제처리하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/pieces/Piece.java:
```diff
@@ -1,95 +1,163 @@
+        return type == piece.type && color == piece.getColor();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(color, type);
+    }
```
혹시 `hashCode`를 오버라이딩 해주신 이유를 알고계실까요~?

kjk402 replied:
equals 로 비교시에 객체가 같다면 hashCode 에서도 같아야 한다고 알고있습니다.
equals 에서 비교값을 color 와 type으로 해주었으니 hashCode 역시 color와 type이 같다면 두 객체가 동일하다라는 판단을 할 수 있도록 오버라이딩 한 것입니다.

ksundong replied:
그게 왜 그런지 조금 더 깊이 고민해봐주세요.
`hashCode`는 어디에 사용되는 메서드일까요?

kjk402 replied:
equals 와 hashCode 모두  object 클래스에서 사용되며 equals는 메모리 주소를 사용하여 객체를 비교합니다. 두 객체가 같다는 의미는 메모리주소와 해시코드가 같다는 의미로 hashCode도 재정의 해줘야한다고 까지 확인했습니다.

ksundong replied:
제가 의도한 것과 조금은 다른 접근이군요. `hashCode` 메서드가 어디에 사용되는지는 [이 글](https://www.geeksforgeeks.org/equals-hashcode-methods-java/)을 읽어보세요.

ksundong replied:
그리고 position이 달라도 같은 것인지도 한 번 생각해볼만 할 것 같아요!
제가 예전에 입사 과제로 받았던 문제와 유사한 문젠데요, 한 번 생각해보세요!

ksundong commented on src/test/test/net/kjk402/chess/chess/BoardTest.java:
```diff
@@ -1,70 +1,144 @@
+                () -> assertThat(whiteKing).isEqualTo(board.findPiece(e3))
         );
+
+        System.out.println(board.showBoard());
```
굳이 출력해줄 필요는 없을 것 같아요~

kjk402 replied:
네 해당 코드 제거해주었습니다.

ksundong commented on src/test/test/net/kjk402/chess/chess/BoardTest.java:
```diff
@@ -1,70 +1,144 @@
+        board.initializeEmpty();
+        String b5 = "b5";
+        Piece blackRook = Piece.createBlackRook(new Position(b5));
+        board.move(b5, blackRook);
```
`move()`를 해주시는 이유는 뭔가요?

kjk402 replied:
앞선 피드백에서 말씀드렸듯이 제가 생각한 move메소드의 역할은 setPiece와 비슷하여 기물 배치용으로 사용되었습니다.

ksundong replied:
그러면, move 할 때, Position을 넣어주는 것이 조금 더 자연스러운 흐름일 것 같은데 어떻게 생각하세요?

ksundong commented on src/test/test/net/kjk402/chess/chess/BoardTest.java:
```diff
@@ -1,70 +1,144 @@
+                () -> assertThat(15.0).isEqualTo(board.caculcatePoint(Piece.Color.BLACK)),
+                () -> assertThat(7.0).isEqualTo(board.caculcatePoint(Piece.Color.WHITE))
+        );
+        System.out.println(board.showBoard());
```
여기도 마찬가지로 출력 코드는 필요 없을 것 같아요.

kjk402 replied:
f492713 에서 삭제해주었습니다.

ksundong replied:
여기에 첨언 한 번 드리자면, 출력을 확인하는 용도로 사용하셔도 무방합니다.
그럴 의도였다면, 굳이 지우지 않으셨어도 됐을 것 같아요.

ksundong commented on src/test/test/net/kjk402/chess/chess/BoardTest.java:
```diff
@@ -1,70 +1,144 @@
+        System.out.println(board.showBoard());
+    }
+
+    private void addPiece(String position, Piece piece) {
+        board.move(position, piece);
+    }
```
굳이 이런 메서드를 정의해줄 필요가 있을까요?

kjk402 replied:
메서드를 활용했을 때 코드 길이가 줄어드는 것이나 가독성이 향상되는 것이 아니기 때문에 해당 메서드 삭제하였습니다.

ksundong commented on src/test/test/net/kjk402/chess/chess/BoardTest.java:
```diff
@@ -1,70 +1,144 @@
+        );
+    }
+
+    void setPiece(List<Piece> list) {
```
`set`을 너무 많이 사용하시는데, `set`은 사용하지 않고 네이밍하는 습관을 들여주세요.

kjk402 replied:
네 알겠습니다.

그렇다면 네이밍에서 get을 사용하는 것 역시 지양해야 할까요?

ksundong replied:
네, `get`, `set` 사용은 Java Beans 규약을 준수하는 정도로 저는 사용하고 있습니다.

ksundong commented on src/test/test/net/kjk402/chess/pieces/PieceTest.java:
```diff
@@ -7,29 +7,38 @@
+
     @Test
     @DisplayName("모든 말을 색상별로 생성하여 확인한다.")
     void create_allPieces() {
```
어디는 snake case고 어디는 camel case를 사용하는 것은 일관적이지 않습니다.
가급적이면 case는 맞춰서 작성해주세요.

ksundong replied:
그리고 해당 테스트 실패하는데, 저만 그런걸까요?

kjk402 replied:
db8ecd2 에서 해당 테스트 정상 동작되도록 수정하였습니다.

ksundong commented on src/test/test/net/kjk402/chess/pieces/PieceTest.java:
```diff
@@ -7,29 +7,38 @@
         );
     }
 
+    @Test
+    public void getRepresentationPerPiece() throws Exception {
+        assertThat(Piece.Type.PAWN.getWhiteRepresentation()).isEqualTo(Piece.Type.PAWN.getWhiteRepresentation());
+        assertThat(Piece.Type.PAWN.getBlackRepresentation()).isEqualTo(Piece.Type.PAWN.getBlackRepresentation());
+    }
```
이 테스트는 하는 동작은 명확한데, 의도한 바가 무엇인지 잘 모르겠네요.

kjk402 replied:
이전 미션에서 사용했던 테스트입니다만, 삭제후 다시 생성하는 과정에서 오탈이 있었던 것 같습니다.
getRepresentation 을 하였을 때 올바른 char 값이 맞는 지 확인하는 테스트입니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,102 @@
+        Position tempPosition = new Position(position);
+        rankList.get(tempPosition.getFile()).move(tempPosition.getRank(), piece);
+        Rank rank = new Rank();
+        rank.addPiece(piece);
```
이 메서드는 기존에 있던 piece가 해당 list에서 제거되고, 다른 list에 추가되는 것이 맞나요?
코드를 봐서는 그렇지 않다는 생각이 듭니다.

kjk402 replied:
말씀하신바와 같은 동작을 구현하기 위해 사용됐으나 이번 미션의 move는 기물이 움직이는 것을 구현하기보다는 특덩 위치에 배치만을 위해 사용되는 것같습니다. 
하여 이번 미션에서는 addPiece부분이 필요없다고 판단되어 삭제처리하였습니다.

ksundong replied:
음 저라면 그냥 `move`는 구현했을 것 같아요.
이건 개인별 의견차가 있을 수 있는 부분이겠네요.

ksundong reviewed (changes requested):
안녕하세요! Bat!! 리뷰어 Dion입니다. 5단계는 요구사항도 복잡하고, 해야할 것도 많아서 많이 고생하셨을 것 같습니다.
요구사항은 대부분 잘 챙겨주셨다고 생각되고, 코드를 리뷰하면서 드는 생각들과 한 번 공부해보셨으면 좋겠다고 생각되는 부분에 코멘트를 작성했습니다.

## 리뷰 내용 요약

- 항상 다형성이 적용된 요소를 사용했을 때에는 최대한 다형성을 이용할 수 있는 방향을 생각해보세요.
- 중복되는 부분이 존재하면, 추출을 시도해봐주세요.
- 네이밍은 어렵지만, 한 가지 제약을 드리자면, `set`을 최대한 아껴서 사용해보시고, `temp`는 정말 임시적인 값에만 사용해주세요.
- 어떤 클래스가 어떤 메서드를 가져야 할 지 조금만 더 생각해주세요.
- 테스트 코드에는 필요하다면 출력을 보여줄 수 있겠지만, 굳이 그렇지 않다면, 테스트 코드에서 출력할 필요가 없는 것 같아요.
- 커밋 메시지 제목에는 온점(.)을 찍지 말아주세요.
- 커밋 메시지는 영어면 영어, 한글이면 한글로 확실하게 작성해주세요.

제가 생각했을 때, 조금 아쉬운 부분과 제 생각에 요구사항을 충족하지 못할 것 같은 부분이 있어 변경요청을 드립니다.
궁금하시거나, 이해가 안되는 부분, 도움이 필요한 부분은 언제든 코멘트 남겨주세요!
고생하셨습니다!

ksundong commented on src/test/test/net/kjk402/chess/pieces/PieceTest.java:
```diff
@@ -7,29 +7,38 @@
-        verifyPiece(Piece.createBlackQueen(), Piece.BLACK, Piece.BLACK_QUEEN_REPRESENTATION);
-        verifyPiece(Piece.createWhiteKing(), Piece.WHITE, Piece.WHITE_KING_REPRESENTATION);
-        verifyPiece(Piece.createBlackKing(), Piece.BLACK, Piece.BLACK_KING_REPRESENTATION);
+        verifyPiece(Piece.createWhitePawn(position), Piece.Color.WHITE, Piece.Type.PAWN);
+        verifyPiece(Piece.createBlackPawn(position), Piece.Color.BLACK, Piece.Type.PAWN);
+        verifyPiece(Piece.createWhiteRook(position), Piece.Color.WHITE, Piece.Type.ROOK);
+        verifyPiece(Piece.createBlackRook(position), Piece.Color.BLACK, Piece.Type.ROOK);
+        verifyPiece(Piece.createWhiteKnight(position), Piece.Color.WHITE, Piece.Type.KNIGHT);
+        verifyPiece(Piece.createBlackKnight(position), Piece.Color.BLACK, Piece.Type.KNIGHT);
+        verifyPiece(Piece.createWhiteBishop(position), Piece.Color.WHITE, Piece.Type.BISHOP);
+        verifyPiece(Piece.createBlackBishop(position), Piece.Color.BLACK, Piece.Type.BISHOP);
+        verifyPiece(Piece.createWhiteQueen(position), Piece.Color.WHITE, Piece.Type.QUEEN);
+        verifyPiece(Piece.createBlackQueen(position), Piece.Color.BLACK, Piece.Type.QUEEN);
+        verifyPiece(Piece.createWhiteKing(position), Piece.Color.WHITE, Piece.Type.KING);
+        verifyPiece(Piece.createBlackKing(position), Piece.Color.BLACK, Piece.Type.KING);
```
여기도 assertAll로 묶어줘도 괜찮을 것 같네요.

kjk402 replied:
해당 피드백 db8ecd2 
assertAll로 적용시켜줬습니다.

ksundong commented on src/test/test/net/kjk402/chess/pieces/PieceTest.java:
```diff
@@ -7,29 +7,38 @@
     }
 
+    @Test
+    public void getRepresentationPerPiece() throws Exception {
```
`public`과 `Exception`은 필요 없으니 제거해주세요.
`public`이 필요없는 이유는 아실 것이라고 생각되어 코멘트하지 않습니다!

kjk402 replied:
해당 피드백 db8ecd2 에서 수정해주었습니다.

kjk402 committed:
PieceTest class 수정

createAllPieces 테스트가 정상 작동 하도록 수정
public 과 Exception 제거.

kjk402 committed:
BlackLine -> BlankLine 오타수정.

kjk402 committed:
Board class 수정

1. ArrayList로 선언된 부분 list로 수정
2. showBoard 메소드의 중복제거
3. findPiece 메소드의 불필요한 괄호 제거, tempPosition ->
   targetPosition으로 수정
4. pieceSort 메소드의 list 명 pieceList로 수정

kjk402 committed:
Rank class 에서 사용되는 메서드명들을  set~ 대신 create~ 로 변경

kjk402 committed:
BlackLine 메소드명 수정

kjk402 committed:
Piece class의 isPawn 메서드 삭제, position을 final로 선언

kjk402 commented:
BoardTest 클래스를 코딩할 때는
미션5 예제 코드를 가져와서 테스트케이스 몇가지 추가 후 정상적으로 테스트 통과가 되는 지만 신경쓰면서 코딩하여 세부사항을 생각하지 못하였습니다. 그러하여 남겨주신 BoardTest 관련 피드백들은 답변을 못하였습니다. 죄송합니다. 다시 고민해보도록하겠습니다. 

그 외 주신 피드백들은 참고하여 코드 수정하였습니다. 

꼼꼼히 리뷰해주셔서 감사드립니다.

ksundong commented:
답변이 몇몇 부분은 빠진 것이 있는데, 궁금했던 부분이어서 확인해주시면 감사하겠습니다~
추가적인 리뷰는 답변을 받고 해드리겠습니다!

kjk402 committed:
getPoint, getRepresentationByRank 메소드 Rank class로 이동

kjk402 committed:
Position final 선언 기본으로 원복.

kjk402 committed:
BoardTest class에서 체스판 출력해주는 코드 삭제

kjk402 committed:
BoardTest class에서 addPiece 메소드 삭제

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
+        StringBuilder boardOutput = new StringBuilder();
+        String rankIndex = "abcdefgh";
+        for (int i = 0; i < BOARD_SIZE; i++) {
+            int boardSizeMinusI = BOARD_SIZE - i;
```
의미는 확 들어오긴 하는데, 좋은 네이밍은 아닌 것 같아요.
저도 딱히 좋은 네이밍이 떠오르지는 않네요. ㅠㅠ

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
-            blackPiecesList.add(piece);
+    public String showBoard() {
+        StringBuilder boardOutput = new StringBuilder();
+        String rankIndex = "abcdefgh";
```
`final`을 써줘도 좋을 것 같네요.

kjk402 replied:
넵 final 붙여줬습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
+        String rankIndex = "abcdefgh";
+        for (int i = 0; i < BOARD_SIZE; i++) {
+            int boardSizeMinusI = BOARD_SIZE - i;
+            boardOutput.append(appendNewLine(Rank.getRepresentationByRank(rankList.get(boardSizeMinusI - 1)) + " " + (boardSizeMinusI)));
```
불필요한 괄호는 제거해주세요.

kjk402 replied:
제거했습니다.~

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
 
-    public Piece findBlackPiece(int idx) {
-        return blackPiecesList.get(idx);
+    public void move(String position, Piece piece) {
```
movePiece가 더 정확하겠죠?

kjk402 replied:
5e81538 에서 movePiece로 변경해주었습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
     }
 
-    public void initialize() {
+    public double caculcatePoint(Color color) {
```
```suggestion
    public double calculatePoint(Color color) {
```

kjk402 replied:
5e81538 에서 철자 수정하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Board.java:
```diff
@@ -1,131 +1,83 @@
+        double pointSum = 0.0;
         for (int i = 0; i < BOARD_SIZE; i++) {
-            addWhitePawn(new Piece(Piece.WHITE, Piece.WHITE_PAWN_REPRESENTATION));
-            addBlackPawn(new Piece(Piece.BLACK, Piece.BLACK_PAWN_REPRESENTATION));
+            pointSum += Rank.getPoint(rankList.get(i), color);
         }
```
```java
for (rank : rankList) {
    pointSum += rank.getPoint(color);
}
```

이런식으로 설계해보는건 어떤가요?

kjk402 replied:
5e81538에서 반영하였습니다.
코드가 훨씬 간결해진 것같습니다. 감사합니다.

ksundong replied:
여기에 더한다면 stream을 사용해볼 수도 있을 것 같아요.
요건 선택입니다~

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.List;
+
+public class Rank {
+    private List<Piece> pieceList = new ArrayList<>(Board.BOARD_SIZE);
```
혹시 이렇게 선언해주면 `ArrayList`는 어떻게 생성되는지 알고 계신가요?
알고계시면 그냥 넘어가셔도 됩니다~

kjk402 replied:
다형성을 위해 업캐스팅을 하는 것이라고만 알고있습니다.
관련내용 자세히 공부해보겠습니다.

ksundong replied:
음 `ArrayList`의 생성자를 의미한거에요!

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,107 @@
+public class Rank {
+    private List<Piece> pieceList = new ArrayList<>(Board.BOARD_SIZE);
+
+    public void addPiece(Piece piece) {
+        pieceList.add(piece);
+    }
```
pieceList의 크기를 측정해서 더이상 add를 할 수 없게하는 제약도 주면 좋을 것 같네요!

kjk402 replied:
2e12034 에서 보드 사이즈보다 작을 때만 넣어주는 것으로 변경하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,107 @@
+        pieceList.set(position, piece);
+    }
+
+    public static double getPoint(Rank rank, Color color) {
+        double sum = 0.0;
+        for (Piece piece : rank.getPieceList()) {
+            if (piece.getColor() == color) {
+                sum += piece.getType().getDefaultPoint();
+            }
+        }
+        return sum;
+    }
+
+    public static String getRepresentationByRank(Rank rank) {
+        StringBuilder sb = new StringBuilder();
+        for (Piece piece : rank.getPieceList()) {
+            sb.append(piece.getRepresentation());
+        }
+        return sb.toString();
+    }
```
`static`을 빼주세요.
어떤 것이 클래스 레벨이어야 하고, 어떤 것이 인스턴스 레벨인지도 꼭 고민해봐주세요.

ksundong replied:
이 부분은 왜 `static`을 빼야하는지 공부해보시고 답변해주세요!

kjk402 replied:
인스턴스 메소드로 수정하였습니다.

인스턴스 변수를 직접 접근 할 수 있도록 하기위해 static이 붙은 클래스 메소드가 아닌 인스턴스 메소드를 사용해야 하기때문에? 인가요.

ksundong replied:
아뇨 객체지향적으로 접근해보자면, 로직을 static으로 구현하는 것은 지양해야합니다.
그 측면에서 보면 Rank를 가지고 로직이 동작하는데 static이어서 Rank를 인자로 받는 구조가 생겼습니다. 이런 부분은 조심해주세요~

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,107 @@
+        return pieceList.get(rankIndex);
+    }
+
+    public static Rank createWhitePiecesRank(int index) {
+        Rank rank = new Rank();
+        rank.addPiece(Piece.createWhiteRook(new Position(0, index)));
+        rank.addPiece(Piece.createWhiteKnight(new Position(1, index)));
+        rank.addPiece(Piece.createWhiteBishop(new Position(2, index)));
+        rank.addPiece(Piece.createWhiteQueen(new Position(3, index)));
+        rank.addPiece(Piece.createWhiteKing(new Position(4, index)));
+        rank.addPiece(Piece.createWhiteBishop(new Position(5, index)));
+        rank.addPiece(Piece.createWhiteKnight(new Position(6, index)));
+        rank.addPiece(Piece.createWhiteRook(new Position(7, index)));
+        return rank;
+    }
```
`index`말고 더 좋은 네이밍을 고민해봐주세요.

kjk402 replied:
2e12034 rankIndex 로 수정하였습니다.

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public static Rank createWhitePawnsRank(int index) {
+        Rank rank = new Rank();
```
이렇게 팩토리 메서드를 사용할 것이라면, 굳이 기본 생성자는 외부에 노출 시킬 필요는 없을 것 같아요.

ksundong replied:
또한 여기도 `index`라는 네이밍 보다 좋은 네이밍을 고민해봐주세요.

kjk402 replied:
2e12034 에서 private로 변경하였습니다.

kjk402 replied:
네이밍 변경하였습니다.

기본생성자 노출을 막기위해선 Board 클래스에 있는 initialize 메소드를 Rank 클래스로 옮기는 것일 좋을까요?

ksundong replied:
private으로 기본생성자를 만들면 컴파일 에러가 발생하나요?

kjk402 replied:
Board 클래스에서 showBoard 메서드가 기본생성자 호출을 해서 에러 발생했습니다.

fb39340 에서 Rank 기본 생성자 private로 변경, showBoard 에러 발생하지 않게 수정하였습니다.

ksundong reviewed (changes requested):
안녕하세요! Bat!! 변경요청된 부분을 수정해주시느라 고생하셨습니다.
다음 단계를 진행하셔도 괜찮을 것 같지만, 조금 더 고민할 부분이 남아있는 것 같아서 변경 요청 드립니다!!

kjk402 committed:
BoardTest class에서 move 테스트 수정

move 할 때 position 지정

kjk402 committed:
Board class 수정

1. move 메소드명 movePiece로 변경
2. 계산 메소드 네이밍 철자 수정

kjk402 committed:
Rank class index명 rankIndex로 변경, 클래스 메소드 인스턴스 메소드로
변경.

ksundong commented on src/main/java/net/kjk402/chess/chess/Rank.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    private List<Piece> getPieceList() {
+        return pieceList;
+    }
```
얘는 왜 감춰두신걸까요?

kjk402 replied:
점수계산과 체스판 출력할 때 사용되던 것인데 Rank 클래스로 옮기면서 같이 private 처리하였습니다.

ksundong replied:
그렇군요!

ksundong reviewed (approved):
수정해주시느라 고생하셨습니다.
리뷰가 늦어져서 죄송해요~

kjk402 committed:
Rank class 기본생성자 private로 변경
